File: engine/src/BoxConstraints.ts
import Size from './Size'

interface IBoxConstraints {
	minWidth: number
	maxWidth: number
	minHeight: number
	maxHeight: number
}

class BoxConstraints implements IBoxConstraints{
	constructor(data: IBoxConstraints) {
		Object.assign(this, data)
	}

	minWidth: number
	maxWidth: number
	minHeight: number
	maxHeight: number

	makeLoose(): BoxConstraints {
		return new BoxConstraints({
			minWidth: 0,
			maxWidth: this.maxWidth,
			minHeight: 0,
			maxHeight: this.maxHeight
		})
	}

	static isTight(constraints: BoxConstraints) {
		return (
			constraints.minWidth === constraints.maxWidth &&
			constraints.minHeight === constraints.maxHeight
		)
	}

	static fromSize(size: Size) {
		return new BoxConstraints({
			minWidth: size.width,
			maxWidth: size.width,
			minHeight: size.height,
			maxHeight: size.height
		})
	}
}

export default BoxConstraints

File: core/src/box_constraints.cpp
#include "box_constraints.hpp"

namespace aardvark {

BoxConstraints BoxConstraints::make_loose() {
    return BoxConstraints{
        0,          // min_width
        max_width,  // max_width
        0,          // min_height
        max_height  // max_height
    };
};

bool BoxConstraints::is_tight() {
    return min_width == max_width && min_height == max_height;
};

Size BoxConstraints::max_size() { return Size{max_width, max_height}; }

BoxConstraints BoxConstraints::from_size(Size size, bool tight) {
    return BoxConstraints{
        tight ? size.width : 0,   // min_width
        size.width,               // max_width
        tight ? size.height : 0,  // min_height
        size.height               // max_height
    };
};

}  // namespace aardvark
